src/GPUActivityAgent.cpp

```diff
@@ -67,26 +71,69 @@ namespace geopm
+
+        std::vector<int> control_domains;
+        control_domains.push_back(m_platform_io.control_domain_type("GPU_CORE_FREQUENCY_MIN_CONTROL"));
+        control_domains.push_back(m_platform_io.control_domain_type("GPU_CORE_FREQUENCY_MAX_CONTROL"));
+
+        std::vector<int> signal_domains;
+        signal_domains.push_back(m_platform_io.signal_domain_type("GPU_CORE_FREQUENCY_STATUS"));
+        signal_domains.push_back(m_platform_io.signal_domain_type("GPU_CORE_ACTIVITY"));
+        signal_domains.push_back(m_platform_io.signal_domain_type("GPU_UTILIZATION"));
+
+        // We'll use the coarsest granularity supported by any of the controls or signals except Energy
+        // i.e. GPU if one control supports GPU and another supports GPU_CHIP
+        int agent_domain = std::min(*std::min_element(std::begin(control_domains), std::end(control_domains)),
+                                    *std::min_element(std::begin(signal_domains), std::end(signal_domains)));
+
+#ifdef GEOPM_DEBUG
+        int max_agent_domain = std::max(*std::max_element(std::begin(control_domains), std::end(control_domains)),
+                                    *std::max_element(std::begin(signal_domains), std::end(signal_domains)));
+
+        if (agent_domain != max_agent_domain) {
+            throw Exception("GPUActivityAgent::" + std::string(__func__) +
+                            "(): Required signals and controls do not all exist at the same domain.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+#endif
+
+        if (agent_domain != GEOPM_DOMAIN_GPU && agent_domain != GEOPM_DOMAIN_GPU_CHIP) {
+            throw Exception("GPUActivityAgent::" + std::string(__func__) +
+                            "(): Required signals and controls do not exist at the " +
+                            "GPU or GPU_CHIP domain!", GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        m_agent_domain_count = m_platform_topo.num_domain(agent_domain);
+
+        for (int domain_idx = 0; domain_idx < m_agent_domain_count; ++domain_idx) {
             m_gpu_core_activity.push_back({m_platform_io.push_signal("GPU_CORE_ACTIVITY",
-                                              GEOPM_DOMAIN_GPU,
-                                              domain_idx), NAN});
+                                           agent_domain,
+                                           domain_idx), NAN});
             m_gpu_utilization.push_back({m_platform_io.push_signal("GPU_UTILIZATION",
-                                         GEOPM_DOMAIN_GPU,
+                                         agent_domain,
                                          domain_idx), NAN});
-            m_gpu_energy.push_back({m_platform_io.push_signal("GPU_ENERGY",
-                                    GEOPM_DOMAIN_GPU,
-                                    domain_idx), NAN});
+        }
 
+        for (int domain_idx = 0; domain_idx < m_agent_domain_count; ++domain_idx) {
```

Comment:
It seems like you can merge this ``for`` loop with the one above, since they iterate over the same range and it doesn't seem like the instructions executed here need to happen separately or after the prior ones.